clang/lib/Sema/SemaTypeTraits.cpp

```diff
@@ -2595,6 +2597,97 @@ static void DiagnoseNonStandardLayoutReason(Sema &SemaRef, SourceLocation Loc,
   SemaRef.Diag(D->getLocation(), diag::note_defined_here) << D;
 }
 
+static void DiagnoseNonAggregateReason(Sema &SemaRef, SourceLocation Loc,
+                                       const CXXRecordDecl *D) {
+  for (const CXXConstructorDecl *Ctor : D->ctors()) {
+    if (Ctor->isUserProvided())
+      SemaRef.Diag(Loc, diag::note_unsatisfied_trait_reason)
+          << diag::TraitNotSatisfiedReason::UserDeclaredCtr;
+    if (Ctor->isInheritingConstructor())
+      SemaRef.Diag(Loc, diag::note_unsatisfied_trait_reason)
+          << diag::TraitNotSatisfiedReason::InheritedCtr;
+  }
+
+  bool HasInherited = llvm::any_of(D->decls(), [](auto const *Sub) {
+    bool Result = false;
+    if (auto *UD = dyn_cast<UsingDecl>(Sub)) {
+      Result = llvm::any_of(UD->shadows(), [](auto const &I) {
+        return isa<ConstructorUsingShadowDecl>(I);
+      });
+    }
+    return isa<ConstructorUsingShadowDecl>(Sub) || Result;
+  });
+
+  if (HasInherited) {
+    SemaRef.Diag(Loc, diag::note_unsatisfied_trait_reason)
+        << diag::TraitNotSatisfiedReason::InheritedCtr;
+  }
```

Comment:
The loop above should be sufficient, right? Why do we  check for inherited constructors twice?